feat(sdk): ability to enforce the destination collection to save to and hide the collection picker

heypoom opened the pull request:
Closes https://github.com/metabase/metabase/issues/49193

### Description

Provides the ability to use `saveOptions.saveToCollectionId` to save the collection to a specified collection id, while hiding the collection picker from the save modal.

### How to verify

- Go to Storybook: http://localhost:6006/?path=/story/embeddingsdk-interactivequestion--default
- Set `isSaveEnabled` to `true` and `saveToCollectionId` to a collection id of choice
- Click on Save
- Click on "Save as new question"
- The collection picker should not be visible now
- Click on Save
- The network request should indicate that you are indeed saving to that `collectionId`

### Checklist

- [x] Tests have been added/updated to cover changes in this PR - E2E test

heypoom committed:
add ability to hide the collection picker

heypoom committed:
add save options

heypoom committed:
make sure collectionId is fresh

heypoom committed:
update sdk question form component

heypoom committed:
make collectionId accept null

heypoom committed:
apply saveOptions to sdk components

heypoom committed:
revert to use initialCollectionId as context name

heypoom committed:
target the correct collection id

heypoom committed:
add e2e tests

heypoom committed:
update e2e tests

heypoom committed:
move into within

heypoom committed:
update test

heypoom committed:
remove collectionId from default args

heypoom commented on frontend/src/metabase/components/SaveQuestionForm/SaveQuestionForm.tsx:
```diff
@@ -54,11 +59,13 @@ export const SaveQuestionForm = ({
-            title={t`Which collection should this go in?`}
-            zIndex={DEFAULT_MODAL_Z_INDEX + 1}
-          />
+          {withCollectionPicker && (
```
Make it possible to hide the collection picker entirely.

heypoom commented on frontend/src/metabase/components/SaveQuestionForm/context.tsx:
```diff
@@ -62,14 +65,34 @@ export const SaveQuestionProvider = ({
+      getInitialValues(
+        originalQuestion,
+        question,
+        initialCollectionId ?? defaultCollectionId,
```
If `initialCollectionId` is passed in, we use that as the initial collection to save to. This mainly applies when the collection picker is visible.

heypoom commented on frontend/src/metabase/components/SaveQuestionForm/context.tsx:
```diff
@@ -53,6 +54,8 @@ export const SaveQuestionProvider = ({
   onCreate,
   onSave,
   multiStep = false,
+  withCollectionPicker = true,
```
Collection Picker should always be present unless the SDK user explicitly disables it in their `saveOptions`. This is not applicable in the main Metabase app currently.

heypoom committed:
Merge branch 'master' into interactive-question-restrict-save-destination

heypoom commented:
PS. I'm not sure who owns the `SaveQuestionForm`, so I'll request review from AdminWebapp for that part for now. I could be completely wrong here.

heypoom committed:
revert e2e of hide columns

heypoom commented on e2e/test/scenarios/embedding-sdk/interactive-question.cy.spec.ts:
```diff
@@ -107,39 +94,51 @@ describeSDK("scenarios > embedding-sdk > interactive-question", () => {
+    tableInteractive().findByText("Product ID").should("be.visible");
 
-    tableHeaderClick("Max of Quantity");
+    tableHeaderClick("Product ID");
```
update: reverting this for now as the runs now pass. will run it through stress test and see if it passes as is. not sure if something is wrong with my local e2e setup...

heypoom replied:
@npretto I rolled back this change now, FYI. Will address it in a separate PR if it arises again locally.

npretto replied:
As a general rule I think it's better if we always do a separate PR for fixing tests, even just for a better git history/git blame

npretto reviewed (commented):
Are we sure this is the correct external API?
Seeing the issue title I was expecting that it was possible to provide more than 1 collectionId.

If we want to limit to only one, should w provide two different params for when we want to set the initial collection id in the collection picker, and for when we want to limit saving to one?
Currently the id param is doing two different things conditional to another prop

heypoom commented:
> Seeing the issue title I was expecting that it was possible to provide more than 1 collectionId.

@npretto I think the issue title is a bit misleading then. For the time being, they want the ability to disable the collection selection so the end user don't need to choose where to save to:

> They don't need to choose where to save to. Just disabling the selection would be enough for the customer at the moment.

> Should we provide two different params for when we want to set the initial collection id in the collection picker, and for when we want to limit saving to one?

That makes sense. Perhaps we can use `saveOptions.saveToCollectionId` for the ability to save directly to a collection, to be explicit. I wonder if we should disable the collection picker when `saveToCollectionId` is provided as that would determine where to save to - does that make sense?

heypoom commented on e2e/test/scenarios/embedding-sdk/helpers/save-interactive-question-e2e-helpers.ts:
```diff
@@ -0,0 +1,80 @@
+  cy.intercept("POST", "/api/card").as("createCard");
+
+  cy.findAllByTestId("cell-data").last().click();
+  popover().findByText("See these Orders").click();
```
@npretto Good point, I should make this parametric then.

heypoom committed:
make e2e helper more generic

heypoom committed:
use saveToCollectionId prop to save to a specific collection

heypoom committed:
update e2e tests

heypoom commented:
@npretto I updated the API to be more straightforward now, with `saveToCollectionId` to save directly to a collection and thus bypassing the collection picker. Thanks!

PS. I didn't implement `initialCollectionId` for pre-filling the initial collection in the save modal, as it's out-of-scope for the issue/PR.

heypoom committed:
Merge branch 'master' into interactive-question-restrict-save-destination

WiNloSt commented:
@heypoom can you update the README? Since we modify the component API.

WiNloSt commented on frontend/src/metabase/components/SaveQuestionForm/util.ts:
```diff
@@ -24,38 +28,53 @@ const updateQuestion = async (
 
-  const collectionId = canonicalCollectionId(details.collection_id);
+  // `saveToCollectionId` is used to override the target collection of the question,
+  // this is mainly used for the embedding sdk.
```
```suggestion
  // this is mainly used for the embedding SDK.
```

npretto commented:
> @heypoom can you update the README? Since we modify the component API.

I think we'll "nuke" the readme with https://github.com/metabase/metabase/pull/49253 and move the docs there, but yeah I agree that we should try to keep the docs and code in sync

WiNloSt commented on frontend/src/metabase/components/SaveQuestionForm/util.ts:
```diff
@@ -24,38 +28,53 @@ const updateQuestion = async (
-  const collectionId = canonicalCollectionId(details.collection_id);
+  // `saveToCollectionId` is used to override the target collection of the question,
+  // this is mainly used for the embedding sdk.
+  const collectionId = isNullOrUndefined(saveToCollectionId)
```
I'd love to see you keep `canonicalCollectionId` here since you keep the variable name the same, but you move `canonicalCollectionId` somewhere else.

This feels like the meaning of this variable already changed, then we might need to change the name.

The nuance of `collectionId` in this version isn't a canonical collection ID anymore.

heypoom replied:
cleaned up in https://github.com/metabase/metabase/pull/49251/commits/c60cf9a62798a62bb0b734580b5fd5baac01f134

WiNloSt reviewed (commented):
I'm out of juice now. I'll continue tomorrow, but here's my general feedback.

WiNloSt commented:
> > @heypoom can you update the README? Since we modify the component API.
> 
> I think we'll "nuke" the readme with #49253 and move the docs there, but yeah I agree that we should try to keep the docs and code in sync

Thanks. I lose track of that. It feels like we should prioritize merging the doc PR first, so we don't need to deal with updating the soon-to-be-killed README.md.

heypoom commented on enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/context/InteractiveQuestionProvider.tsx:
```diff
@@ -50,8 +50,9 @@ export const InteractiveQuestionProvider = ({
   onSave,
-  isSaveEnabled = true,
   entityTypeFilter,
+  isSaveEnabled,
```
@WiNloSt this is an artifact from rolling back my changes, as I added a `saveOptions` prop in an earlier version of this PR (https://github.com/metabase/metabase/pull/49251/commits/8fdeb95a9ad99258aec8ab6947cb87ae0a0994f8), but later reverted it. I'll cleanup this diff then.

heypoom replied:
cleaned up in https://github.com/metabase/metabase/pull/49251/commits/c60cf9a62798a62bb0b734580b5fd5baac01f134

jeff-bruemmer commented:
> It feels like we should prioritize merging the doc PR first, so we don't need to deal with updating the soon-to-be-killed README.md.

Merged!

heypoom commented on enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/SdkSaveQuestionForm.tsx:
```diff
@@ -26,7 +26,7 @@ export const SdkSaveQuestionForm = ({ onClose }: SdkSaveQuestionFormProps) => {
       onCreate={onCreate}
       onSave={onSave}
       multiStep={false}
-      initialCollectionId={null}
```
@WiNloSt good point, let me clean that up too.

heypoom replied:
cleaned up in https://github.com/metabase/metabase/pull/49251/commits/c60cf9a62798a62bb0b734580b5fd5baac01f134

heypoom commented on enterprise/frontend/src/embedding-sdk/components/private/QuestionEditor/QuestionEditor.tsx:
```diff
@@ -16,11 +16,11 @@ const QuestionEditorInner = () => {
     originalQuestion,
     onSave,
     onCreate,
+    isSaveEnabled,
```
This is an artifact from rolling back my changes, as I added a saveOptions prop in an earlier version of this PR (https://github.com/metabase/metabase/commit/8fdeb95a9ad99258aec8ab6947cb87ae0a0994f8), but later reverted it. I'll cleanup this diff 👍🏻

heypoom replied:
cleaned up in https://github.com/metabase/metabase/pull/49251/commits/c60cf9a62798a62bb0b734580b5fd5baac01f134

heypoom committed:
update comment formatting

Co-authored-by: Mahatthana (Kelvin) Nomsawadi <me@bboykelvin.dev>

heypoom committed:
clean up leftover changes from an earlier version

heypoom committed:
readme update

heypoom commented on enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/context/types.ts:
```diff
@@ -3,20 +3,24 @@ import type { PropsWithChildren } from "react";
-};
+
+  /** Is the save question button visible? */
+  isSaveEnabled?: boolean;
```
cleaned up in https://github.com/metabase/metabase/pull/49251/commits/c60cf9a62798a62bb0b734580b5fd5baac01f134

heypoom commented:
@WiNloSt I updated the docs in `docs/embedding/sdk` and cleaned up the changes in https://github.com/metabase/metabase/pull/49251/commits/c60cf9a62798a62bb0b734580b5fd5baac01f134. Thanks!

heypoom committed:
Merge branch 'master' into interactive-question-restrict-save-destination

heypoom commented on docs/embedding/sdk/questions.md:
```diff
@@ -69,18 +69,19 @@ export default function App() {
+| customTitle        | `string \| undefined`                              | (optional) Allows a custom title to be displayed instead of the default question title. Only relevant when using the default layout.                                                                                                                                                                               |
+| onBeforeSave       | `() => void`                                       | (optional) A callback function that triggers before saving. Only relevant when `isSaveEnabled = true`.                                                                                                                                                                                                             |
+| onSave             | `() => void`                                       | (optional) A callback function that triggers when a user saves the question. Only relevant when `isSaveEnabled = true`.                                                                                                                                                                                            |
+| saveToCollectionId | `number`                                           | (optional) The target collection to save the question to. This will hide the collection picker from the save modal.                                                                                                                                                                                                |
```
The diffs is a bit dramatic here - the prop name `saveToCollectionId` is long so the table gets expanded.

npretto replied:
The "Hide whitespace changes" thankfully makes the diff a lot better 🚀

heypoom replied:
Added in https://github.com/metabase/metabase/pull/49251/commits/1c74368a069656251ea9f737514142cb2d2a8634

By the way, I think we should actually add a separate section for static and interactive question props. I might do that in a later PR.

WiNloSt replied:
Yeah I figure, maybe we should have a section where the props apply to all question components and maybe props that only exist for a specific component to avoid confusion. Now we're just throwing everything in the same table.

heypoom commented on e2e/test/scenarios/embedding-sdk/interactive-question.cy.spec.ts:
```diff
@@ -107,39 +94,51 @@ describeSDK("scenarios > embedding-sdk > interactive-question", () => {
+    cy.wait("@createCard").then(({ response }) => {
+      expect(response?.statusCode).to.equal(200);
+      expect(response?.body.name).to.equal("Sample Orders 1");
+      expect(response?.body.collection_id).to.equal(null);
```
Ah, that makes sense! My guess is `null` defaults to a sensible default collection for a user? (I could be wrong)

WiNloSt replied:
If a user doesn't have access to the root collection then how can they create a question in the root collection?

AFAIK, we have two ways to represent the root collection one is `"root"` and another one is `null`.

![image](https://github.com/user-attachments/assets/42131464-9e48-4d42-95f6-07f9762ad010)

heypoom replied:
@WiNloSt This is unrelated to this PR, but I just wonder if users without root collection access will get a sensible default collection in the collection picker that they have write access to, or does it always default to the root collection. My guess it that they can see that the collection exist, but can't save to it.

WiNloSt replied:
From my testing, users would be led to the first available collection. Please note that the permission isn't permissive, even if the `root` collection are inaccessible users would still be able to access nested collections if they have permissions for those collections.

Otherwise if all collections are inaccessible, they can only access their personal collection.

heypoom replied:
Oh nice, that's the ideal behaviour then!

WiNloSt commented on enterprise/frontend/src/embedding-sdk/components/private/QuestionEditor/QuestionEditor.tsx:
```diff
@@ -137,19 +137,21 @@ const QuestionEditorInner = () => {
   onSave,
   plugins,
   entityTypeFilter,
+  isSaveEnabled,
```
is the removal of `isSaveEnabled` default value here still valid?

heypoom replied:
@WiNloSt Good catch, I indeed missed that. Fixed in https://github.com/metabase/metabase/pull/49251/commits/79b87dcd2c873a06a91248310c17902b5e5c4893

WiNloSt commented on frontend/src/metabase/components/SaveQuestionForm/types.ts:
```diff
@@ -9,12 +9,37 @@ export type SaveQuestionProps = {
+   *
+   * When this is defined, the collection picker will be hidden and
+   * the question will be saved to this collection.
+   ***/
```
One extra `*` can be dropped 😉 to stay consistent with the rest of the comments

```suggestion
   **/
```

heypoom replied:
Good eye! I didn't notice that

WiNloSt reviewed (approved):
I tested that this works as expected, but please take a look at the failed type check.

heypoom committed:
indicate that saveToCollectionId is only applicable to static questions

heypoom committed:
revert QuestionEditor.isSaveEnabled defaults

heypoom committed:
update comment format

heypoom committed:
remove unused initialCollectionId prop

heypoom committed:
Merge branch 'master' into interactive-question-restrict-save-destination

heypoom committed:
Merge remote-tracking branch 'origin/master' into interactive-question-restrict-save-destination

heypoom committed:
experiment: try to fix failing custom-columns e2e

heypoom commented:
Turns out the `initialCollectionId` is still needed (see commit https://github.com/metabase/metabase/pull/49251/commits/5912e1c1e88b30ba08bb6ad1f8249f2a23b0d399), otherwise the custom-columns-e2e test fails.

heypoom commented:
Nevermind, I don't think the initialCollectionId is the cause here, it's just that custom-columns-ee flaked all of the 8 times I retried the check (really unlucky); I see Kamil has a fix: https://github.com/metabase/metabase/pull/49476
